source/common/stats/tag_extractor_impl.cc

```diff
@@ -169,5 +180,89 @@ bool TagExtractorRe2Impl::extractTag(absl::string_view stat_name, std::vector<Ta
   return false;
 }
 
+TagExtractorTokensImpl::TagExtractorTokensImpl(absl::string_view name, absl::string_view tokens,
+                                               absl::string_view substr)
+    : TagExtractorImplBase(name, tokens, substr), tokens_(absl::StrSplit(tokens, '.')),
+      match_index_(findMatchIndex(tokens_)) {
+  prefix_ = tokens_[0];
+}
+
+uint32_t TagExtractorTokensImpl::findMatchIndex(const std::vector<std::string>& tokens) {
+  for (uint32_t i = 0; i < tokens.size(); ++i) {
+    if (tokens[i] == "$") {
+      return i;
+    }
+  }
+  ASSERT(false, absl::StrCat("did not find match in ", absl::StrJoin(tokens, ".")));
+  return 0;
+}
+
+bool TagExtractorTokensImpl::extractTag(TagExtractionContext& context, std::vector<Tag>& tags,
+                                        IntervalSet<size_t>& remove_characters) const {
+  PERF_OPERATION(perf);
+  const std::vector<absl::string_view>& input_tokens = context.tokens();
+  uint32_t match_input_index = input_tokens.size(), start = 0;
+  if (!extractTagHelper(input_tokens, 0, 0, 0, start, match_input_index)) {
+    PERF_RECORD(perf, "tokens-miss", name_);
+    PERF_TAG_INC(missed_);
+    return false;
+  }
+  const absl::string_view tag_value = input_tokens[match_input_index];
+  uint32_t end = start + tag_value.size();
+  if (match_input_index < (input_tokens.size() - 1)) {
+    ++end; // Remove dot leading to next token
+  } else if (start > 0) {
+    --start;
+  }
+  remove_characters.insert(start, end);
+  addTag(tags) = tag_value;
+
+  PERF_RECORD(perf, "tokens-match", name_);
+  PERF_TAG_INC(matched_);
+  return true;
+}
+
+bool TagExtractorTokensImpl::extractTagHelper(const std::vector<absl::string_view>& input_tokens,
+                                              uint32_t input_index, uint32_t pattern_index,
+                                              uint32_t char_index, uint32_t& start,
+                                              uint32_t& match_input_index) const {
+  for (; input_index < input_tokens.size() && pattern_index < tokens_.size();
+       ++input_index, ++pattern_index) {
+    if (pattern_index == match_index_) {
+      start = char_index;
+      match_input_index = input_index;
+    } else {
+      const absl::string_view expected = tokens_[pattern_index];
+      if (expected == "**") {
+        if (pattern_index == tokens_.size() - 1) {
+          pattern_index = tokens_.size();
+          input_index = input_tokens.size();
+          break;
+        }
+
+        // A "**" in the pattern anywhere except the end means that we must find
+        // a match for the remainder of the pattern anywhere in the the
+        // input. Consider pattern "a.**.b.c.$" and input "a.x.b.b.c.d". We
+        // don't want to just greedily match the "**" against "x" when it should
```

Comment:
I think of it as greedy if I depend on the first token-match I find, even if that isn't a full match. But maybe it depends on your perspective?

Reply:
Yeah, I tend to think of it as a special form of regex. And in regexes greedy match means longest possible string, that is "x.b" in this example. That's why I got confused. Probably "eagerly" could be used here, but again it's a loaded word :)

Anyway the example illustrates the intent very well, so I'd rather drop the adverb to avoid confusions.